### PR TITLE
refactor: introduce State and MutState type classes

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

File: src/State.flix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+        pub def fromState(x: AiState): Option[Bool] = match x {
+            case Value("true") => Some(true)
+            case Value("false") => Some(false)
+            case _ => None
+        }
+    }
+
+    instance State[Int64] {
+        pub def toState(x: Int64): AiState = Value("${x}")
```
Comment (JonathanStarup) [inline src/State.flix:26]: looks good to me. I think if we were very concerned about correctness then I'm not sure that it is guaranteed that `fromString(toString(x)) == x` is always true. I don't think its a law of the type classes but I'm not sure

File: src/State.flix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    use Flixball/Core.AiState;
+    use Flixball/Core.AiState.{Value, Collection};
+
+    pub class State[a] {
+        pub def toState(x: a): AiState
+        pub def fromState(x: AiState): Option[a]
+    }
+
+    pub class MutState[m: Region -> Type] {
+        pub def toState(x: m[r]): AiState \ r
```
Comment (JonathanStarup) [inline src/State.flix:11]: shouldn't these be `Write(r)`? or do we not care yet?

Comment (JonathanStarup) [inline src/State.flix:11]: shouldn't these be `Write(r)`? or do we not care yet?

File: src/AiLibrary.flix [hunk 3]
```diff
@@ -25,18 +25,18 @@ namespace Flixball/AiLibrary {
         use RegionedRandom.nextInt64;
         MoveLogic((_, _) -> state -> region r {
             // use "0" if there's a bad encoding
-            let ran = RegionedRandom.newWithSeed(r, toInt64(state) |> Option.getWithDefault(0i64));
+            let ran = RegionedRandom.newWithSeed(r, fromState(state) |> Option.getWithDefault(0i64));
```
Comment (JonathanStarup) [inline src/AiLibrary.flix:28]: The functions shouldn't really use this directly right? We'd add a function `fixState(ml: GameState -> Id -> s -> (Move, s)): Movelogic with State[s]` for the functional ai's and another one for the mutstate ones

**Review by JonathanStarup — COMMENTED** [review]
Looks nice, just need more QOL functions on top

**Comment by JonathanStarup** [comment]
Oh, and I really think the parsing should retun Option or Result

File: src/AiLibrary.flix (hunk 3 above)
Comment (mlutze) [reply src/AiLibrary.flix:28]: Right. I'll follow up with another PR to use the new function.

File: src/State.flix (hunk 1 above)
Comment (mlutze) [reply src/State.flix:26]: That's a good point. I think as long as we don't rely on it _in general_ we should be ok. It happens to hold for `Int64` and `Bool`, so we can use it for those specifically.

**Comment by mlutze** [comment]
> Oh, and I really think the parsing should retun Option or Result

Isn't that what I'm doing? `fromState` returns an `Option[a]`

File: src/State.flix (hunk 1 above)
Comment (JonathanStarup) [reply src/State.flix:26]: yeah. We could also use the bool instance of `FromString` here just to remove a couple strings

Comment (JonathanStarup) [reply src/State.flix:26]: yeah. We could also use the bool instance of `FromString` here just to remove a couple strings

**Pull request merged**